File: lib/rag/embed.ts
// SiliconFlow (硅基流动) hosts BGE-M3 with an OpenAI-compatible endpoint.
// Docs: https://docs.siliconflow.cn/cn/api-reference/embeddings/create-embeddings
//
// We use the bare fetch API here (not the OpenAI SDK) because we want to
// batch many inputs in one call, and SiliconFlow's batching shape matches
// OpenAI's "input: string | string[]" form cleanly.

export const EMBED_MODEL = process.env.EMBED_MODEL ?? 'BAAI/bge-m3'
export const EMBED_DIM = 1024
const BASE = process.env.SILICONFLOW_BASE_URL ?? 'https://api.siliconflow.cn/v1'

export async function embed(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return []
  const key = process.env.SILICONFLOW_API_KEY
  if (!key) throw new Error('SILICONFLOW_API_KEY not set')

  const res = await fetch(`${BASE}/embeddings`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${key}`,
    },
    body: JSON.stringify({
      model: EMBED_MODEL,
      input: texts,
      encoding_format: 'float',
    }),
  })

  if (!res.ok) {
    const detail = await res.text().catch(() => '')
    throw new Error(`SiliconFlow embed failed: ${res.status} ${detail.slice(0, 300)}`)
  }

  const json = await res.json() as { data: { embedding: number[]; index: number }[] }
  // Re-order by index just in case the API doesn't guarantee it.
  const out: number[][] = new Array(texts.length)
  for (const row of json.data) out[row.index] = row.embedding
  return out
}

// Batch helper for offline embed scripts. SiliconFlow accepts ~64 inputs per call, but
// we batch smaller (16) so a single retry on failure is cheap.
export async function embedBatched(texts: string[], batchSize = 16): Promise<number[][]> {
  const out: number[][] = []
  for (let i = 0; i < texts.length; i += batchSize) {
    const batch = texts.slice(i, i + batchSize)
    const vecs = await embed(batch)
    out.push(...vecs)
  }
  return out
}

File: server/stratsquad/rag/embed.py
"""BGE-M3 embedding via SiliconFlow's OpenAI-compatible endpoint."""
from __future__ import annotations
import os
from typing import Optional
import httpx


EMBED_MODEL = os.getenv("EMBED_MODEL", "BAAI/bge-m3")
EMBED_DIM = 1024
BASE = os.getenv("SILICONFLOW_BASE_URL", "https://api.siliconflow.cn/v1")


async def embed(texts: list[str]) -> list[list[float]]:
    if not texts:
        return []
    key = os.getenv("SILICONFLOW_API_KEY")
    if not key:
        raise RuntimeError("SILICONFLOW_API_KEY not set")

    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.post(
            f"{BASE}/embeddings",
            headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
            json={"model": EMBED_MODEL, "input": texts, "encoding_format": "float"},
        )
        res.raise_for_status()
        data = res.json()["data"]

    out: list[Optional[list[float]]] = [None] * len(texts)
    for row in data:
        out[row["index"]] = row["embedding"]
    return [v for v in out if v is not None]


async def embed_batched(texts: list[str], batch_size: int = 16) -> list[list[float]]:
    out: list[list[float]] = []
    for i in range(0, len(texts), batch_size):
        out.extend(await embed(texts[i:i + batch_size]))
    return out
